### ISSUE TITLE
DI - Injecting IEventStoreConnection

### ISSUE THREAD

**Issue by rm2k** [EventStore/ClientAPI.NetCore#7]
I guess this is more of a question.

I'd like to inject `IEventStoreConnection` into my app,  so that my classes become honest with their dependencies. But I find it challenging since creating the connection is 2 step process:

    var con = EventStoreConnection.Create(...)
    con.ConnectAsync()

I would do something like this this in my `Startup.cs`:

    services.AddScoped(p => 
    {
       // build conString
       return EventStoreConnection.Create(conString);
    });

injected into:

    public class GetEventStore : IEventStore
    {
      private readonly IEventStoreConnection _con;
      public GetEventStore(IEventStoreConnection con)
      {
        _con = con;
      }

      public TResult1 CallMe(...) { ... }
      public TResult2 CallMeToo(...) { ...  }
    }
  
My puzzle is where should I `_con.ConnectAsync()` given that i'm allowed to connect once. Is there a way to check connection status? 

I thought of connecting in my `Startup.cs` before returning the connection or in my constructor(lot of conflicting arguments on creating a connection inside a constructor). 

Long story short, what's the correct to inject my connection? Also, Scoped or Singleton?

**Comment by jageall** [comment]
EventstoreConnection is designed to be reused (there is overhead in establishing the connection so it remains live and has options to reconnect on failure) so use singleton. if you are going to use DI, you just add a function that returns a connected connection. you can either call .Wait() on the connectAsync task or register it as a Task<IEventStoreConnection> and take that as a dependency and await the result. (TPL has a fast path check for tasks that already have a result so for the most part it will incur little overhead if it is already connected).

**Comment by rm2k** [comment]
thanks @jageall.

**Comment by rm2k** [comment]
I have registered eventstore connection as a Singleton, but discovered that if the server shutdown while the app is still running the connection get disposed and become unusable as throws `object disposed` exception thereafter even after restarting the server. The only way out is to restart the app.

Is that a normal behavior? Is there a way to regain connection after a server restart?

**Comment by rm2k** [comment]
`MaxReconnections`. my bad.
